### DISCUSSION TITLE
Why is that using multiple workers is slower than using only one worker

### DISCUSSION THREAD

**Discussion by jinluyang** [fastapi/fastapi#10451] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class ModelService(): # a class that do pytorch inference
    def __init__(self):
        load_pytorch_model(...)
    async def detect(self, inputTensor):
        return await do_compute(inputTensor)

class AsyncGunicornApplication(BaseApplication):
    def __init__(self, options=None):
        self.options = options or {}
        self.application = None
        super().__init__()

    def load_config(self):
        config = {key: value for key, value in self.options.items()
                  if key in self.cfg.settings and value is not None}
        for key, value in config.items():
            self.cfg.set(key.lower(), value)

    def load(self):
        if self.application is None:
            self.application = FastAPI()
        hslog = logging.getLogger("high_service")
        ms = ModelService()

        @self.application.post("/")
        async def process(input:str = Form(...)):
            starttime = time.time()
            input = json.loads(input)
            ret = await ms.detect(input)
            log.info("request cost time:{}".format(time.time() - starttime))
            return ret
        return self.application

options = {
          "bind": "0.0.0.0:" + port,
          "workers": workers,
          "worker_class": "uvicorn.workers.UvicornWorker",
          'timeout': 0
      }

AsyncGunicornApplication(options).run()
```

### Description

setting the `workers` to 1 can get much larger throughput (as well as lower latency)than setting it to 3 or bigger. But if i don't define the process function as "async", setting multiple workers help improve the thoughput. What is the reason of that? Should I not use multiple workers(processes) when using async?

### Operating System

Linux

### Operating System Details

The number of CPUs is 92. Using GPU to do the pytorch compute.

### FastAPI Version

0.103.2

### Pydantic Version

1.10.12

### Python Version

3.8

**Comment by jgould22** [comment, marked as answer]
Sounds like it could be a couple things but at the very least the fact that you say removing async from `process` improves throughput implies you are blocking the event loop.

When you change process to `def` instead of `async def` fast api is spawning things to a threadpool.  [Async python is co-operative multitasking](https://en.wikipedia.org/wiki/Cooperative_multitasking) which means you do not want a lot of processing between await calls.

If you have blocking code in your process function then you should not make it async.
